07-23-blog-core-extension (#4981)
Azure/AKS: blog/_posts/2024-05-14-core-extension.md

---

> title: "Introducing Core Kubernetes Extensions for AKS"

Reviewer: I get it we need to distinguish between 'core' and the other generic extensions wrt degree of integration with aks. on the naming, wondering if 'core' conveys that? thoughts on 'native' or other words? (I don't have a strong preference here, just brainstorming on name)

description: "Learn what are core Kubernetes extensions, and how that extends functionality of your AKS clusters"

> date: 2025-05-14 # date is important. future dates will not be published

Reviewer 2: The name of the file doesn't match this date. Not sure if it matters

author: Jane Guo # must match the authors.yml in the _data folder
categories: 
- extensions 

> # - general

Reviewer 3: Ah, if we have to pick a single category this seems reasonable.

> # - general

Author: Agree that it's also relevant to add-ons. I added both to the categories.

# - general
# - operations
# - networking
# - security
# - developer
# - add-ons
# - extensions
---

## Introducing Core Kubernetes Extensions

### What Are Kubernetes Extensions?

> [Kubernetes extensions](https://learn.microsoft.com/en-us/azure/aks/cluster-extensions?tabs=azure-cli) (or cluster extensions) are pre-packaged applications that simplify the installation and lifecycle management of Azure capabilities on Kubernetes clusters. Examples include [Azure Backup](https://learn.microsoft.com/en-us/azure/backup/azure-kubernetes-service-backup-overview), [GitOps (Flux)](https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/conceptual-gitops-flux2), and [Azure Machine Learning](https://learn.microsoft.com/en-us/azure/machine-learning/how-to-attach-kubernetes-anywhere?view=azureml-api-2). Third-party extensions (or Kubernetes apps), such as [Datadog AKS Cluster Extension](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/datadog1591740804488.dd_aks_extension?tab=Overview) and [Isovalent Cilium Enterprise](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/isovalentinc1662143158090.isovalent-cilium-enterprise?tab=Overview), are also available in the Azure Marketplace.

Reviewer 4: ```suggestion
[Kubernetes extensions](https://learn.microsoft.com/en-us/azure/aks/cluster-extensions?tabs=azure-cli) (or cluster extensions) are pre-packaged applications that simplify the installation and lifecycle management of Azure capabilities on Kubernetes clusters. Examples include [Azure Backup](https://learn.microsoft.com/en-us/azure/backup/azure-kubernetes-service-backup-overview), [GitOps (Flux)](https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/conceptual-gitops-flux2), and [Azure Machine Learning](https://learn.microsoft.com/en-us/azure/machine-learning/how-to-attach-kubernetes-anywhere?view=azureml-api-2). Third-party extensions (or Kubernetes apps), such as [Datadog AKS Cluster Extension](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/datadog1591740804488.dd_aks_extension?tab=Overview) and [Isovalent Cilium Enterprise](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/isovalentinc1662143158090.isovalent-cilium-enterprise?tab=Overview), are also available in the [Azure Marketplace](https://azuremarketplace.microsoft.com/en-US/).
```
nit; link to the Azure Marketplace.

> [Kubernetes extensions](https://learn.microsoft.com/en-us/azure/aks/cluster-extensions?tabs=azure-cli) (or cluster extensions) are pre-packaged applications that simplify the installation and lifecycle management of Azure capabilities on Kubernetes clusters. Examples include [Azure Backup](https://learn.microsoft.com/en-us/azure/backup/azure-kubernetes-service-backup-overview), [GitOps (Flux)](https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/conceptual-gitops-flux2), and [Azure Machine Learning](https://learn.microsoft.com/en-us/azure/machine-learning/how-to-attach-kubernetes-anywhere?view=azureml-api-2). Third-party extensions (or Kubernetes apps), such as [Datadog AKS Cluster Extension](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/datadog1591740804488.dd_aks_extension?tab=Overview) and [Isovalent Cilium Enterprise](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/isovalentinc1662143158090.isovalent-cilium-enterprise?tab=Overview), are also available in the Azure Marketplace.

Reviewer: can we remove 'en-us' part throughout doc for better localization support on docs?

### Extensions vs. Add-ons

Extensions and add-ons both enhance AKS functionality but differ in scope and management:

- **Add-ons**: Part of the AKS resource provider (`Microsoft.ContainerService/managedClusters`), limited to AKS, and generally reserved for cluster critical functionality.
- **Extensions**: Managed via the Extension API (`Microsoft.KubernetesConfiguration/extensions`), supporting both AKS and [Arc-enabled Kubernetes environments](https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/overview).

> 

Reviewer: for novice readers, can we add 1-2 lines/links for intro on following -

- ARM namespace, resource type
- link to swagger or REST API docs
- conceptual design of extension (how unlike addons design from past) and how it allows for delivery to both AKS and Arc

>  Core Kubernetes extensions are a type of Kubernetes extensions that provide broader region availability, a native AKS experience, safer version management rules, and enhanced security and efficiency on AKS. Important AKS capabilities can be supported through core Kubernetes extensions while maintaining seamless integration with AKS just like add-ons. This approach also paves the way for more services and applications to be supported across both AKS and [Arc-connected Kubernetes](https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/overview) environments in the future.

Reviewer 3: and/or something like: "all core extensions are extensions. Not all extensions are core extensions"

>  Core Kubernetes extensions are a Kubernetes extension that provides broader region availability, a native AKS experience, safer version management, enhanced security, efficiency, and reliability on AKS. Important AKS capabilities can be supported through core Kubernetes extensions while maintaining seamless integration with AKS just like add-ons. This approach also paves the way for more services and applications to be supported across both AKS and Arc-connected Kubernetes environments in the future.

Reviewer: nit: there's a leading space before Core

Reviewer: On 'This approach also paves the way for more services and applications to be supported across both AKS and Arc-connected Kubernetes environments in the future' part -
- we typically use service for Azure services and the hosted parts. can we clarify/reword on 'services' part, may be with examples? Are we referring to marketplace offerings or more managed offerings from Azure?
- applications - are we referring to marketplace offerings or anything beyond that?
- Arc-connected Kubernetes - nit: it should be 'Azure Arc-enabled Kubernetes', that's the approved name from naming

Reviewer: and one more benefit - more resource freed up on customer nodes, so less cost to customer

Author: reworded as "functionality"

Author: Regarding more resource freed up on customer nodes, do you mean it as a benefit of moving the extension manager is moved to underlay? If so, I've touched upon it in the following content.

#### Broader region availability

> Core Kubernetes extensions are available in all regions across all clouds (Public cloud, Fairfax, Mooncake, and Air Gapped cloud). See the full list at [Azure Arc enabled Kubernetes region support](https://azure.microsoft.com/en-us/explore/global-infrastructure/products-by-region/?products=azure-arc&regions=all).

Reviewer 5: Please use - "will be available"
Both Extensions Platform and BackUp extension are getting ready to be available in Air gapped environment.

> Core Kubernetes extensions are available in all regions across Azure Public cloud, Azure Government and China regions. See the full list of regions at [Azure Arc enabled Kubernetes region support](https://azure.microsoft.com/en-us/explore/global-infrastructure/products-by-region/?products=azure-arc&regions=all).

Reviewer: - nit: there's a '-' missing between Arc and enabled
- should we also add a link to AKS region availability?

#### Native AKS experience

> * Add-ons migrated to core extensions: Some AKS add-ons, such as Container Insights, Azure Managed Prometheus, and App Monitoring, are transitioning to core extensions. Although the underlying technology shifts to extensions, the user experience remains unchanged. The CLI command to enable/disable migrated add-ons are the same as before migration.

Reviewer: - nit: need to add Azure Monitor branding for the three experiences mentioned here for their official names (which should be available in docs). App Monitoring I'm guessing is referring to Application Insights? if yes, that should be revised too
- on user experience remains unchanged, can we add that the API remains unchanged, and then CLI/portal/Bicep,...?

    CLI command to enable a core Extension that transitioned from an add-on:
    ```bash

>     az aks create/update --resource-group <group> --name <cluster> --enable-<core-extension>

Reviewer 3: ```suggestion
    az aks create/update --resource-group <group> --name <cluster> --enable-<addon>
```

Though it might actually make even more sense to just omit this and say explicitly in the "user experience remains unchanged" section that the CLI commands to enable/disable migrated addons are exactly the same as before migration.

>     az aks create/update --resource-group <group> --name <cluster> --enable/disable <add-on>

Reviewer: it's not going to be just '--enable/disable' params and the scenario name would also be in that parameter no? sharing an example would be good. [This](https://learn.microsoft.com/en-us/azure/defender-for-cloud/defender-for-containers-enable?tabs=k8s-deploy-cli%2Ck8s-deploy-asc%2Ck8s-verify-asc%2Ck8s-remove-arc%2Caks-removeprofile-api&pivots=defender-for-container-aks#use-the-azure-cli-to-deploy-the-defender-sensor) one has an example - 

```
az aks create --enable-defender --resource-group <your-resource-group> --name <your-cluster-name>
```

    ```

> * Shifting Regular Extensions to Core Extensions: Key extensions like Azure Backup are also being shifted to core extensions for better integration with AKS.

Reviewer 3: Maybe instead of "shifted" here, say they're "graduating" or something like that?

Author: I like graduating!

> * Regular extensions graduated to core extensions: Key extensions like Azure Backup are also being upgrading to core extensions for better integration with AKS.

Reviewer: can we add a link to the existing KubernetesConfiguration extensions list? and add a note on which ones are being upgraded to core extensions right now? and a tentative timeline for when the rest will be upgraded?

> 

Reviewer 4: ```suggestion

#### Creating or Deleting Extensions
```

Think adding a header here could be useful. I was looking for the CLI commands to enable core addons, and took me a second to zero in on this section.

    CLI commands:
    * Create/delete a regular extension:
    ```bash
    az k8s-extension create/delete --extension-type <type> --resource-group <group> --cluster-name <name> --cluster-type <clusterType> --name <extension name>
    ```
    * Create/delete a core extension:
    ```bash

>     az aks extension create/delete --extension-type <type> --resource-group <group> --cluster-name <name> --name <core extension name>

Author: --enable/disable is reserved for core extensions migrated from add-ons to keep the original experience. I see that it may be confusing. I've reorganized this section to make it clearer.

    ```

#### Safer version management rules

> Core extensions follow the same update rules as add-ons: patch versions can be upgraded within a Kubernetes minor version, while major/minor upgrades occur only with Kubernetes minor version updates to avoid introducing breaking changes in a Kubernetes version.

Reviewer 3: May be worth noting that some extensions can bypass this requirement if they've demonstrated specific testing/protection against unannounced breaking changes

> In principle, core extensions follow the same update rules as add-ons: patch versions can be upgraded within a Kubernetes minor version, while major/minor upgrades occur only with Kubernetes minor version updates to avoid introducing breaking changes in a Kubernetes version. Exceptions are allowed if the extension has proven testing and safeguards against unexpected breaking changes in new minor/minor versions.

Reviewer 2: typo: minor/minor?

> In principle, core extensions follow the same update rules as add-ons: patch versions can be upgraded within a Kubernetes minor version, while major/minor upgrades occur only with Kubernetes minor version updates to avoid introducing breaking changes in a Kubernetes version. Exceptions are allowed if the extension has proven testing and safeguards against unexpected breaking changes in new minor/major versions.

Reviewer: +1. I don't think we should be capturing that exceptions line. AFAIK I thought we weren't making exceptions. let's discuss this one internally.

also can we link to that versioning policy note in our documentation?

> In principle, core extensions follow the same update rules as add-ons: patch versions can be upgraded within a Kubernetes minor version, while major/minor upgrades occur only with Kubernetes minor version updates to avoid introducing breaking changes in a Kubernetes version. Exceptions are allowed if the extension has proven testing and safeguards against unexpected breaking changes in new minor/major versions.

Reviewer: in addition to link to docs where this versioning policy is mentioned for addons and extension, I think for the blog we have an opportunity to go deeper on the narrative on why/how we arrived at that policy - by discussing the tradeoff between reliability vs speed of availability of new functionality vs patches/fixes

#### Improved security and efficiency

> ExtensionManager is moved from user node pool to AKS Control Plane, which improved extension security and efficiency.

Reviewer 2: also reliability.

> ExtensionManager is moved from user node pool to AKS Control Plane, which improved extension security, efficiency, and reliability.

Reviewer 6: Some small edits"ExtensionManager is moved from _the_ user node pool to _the_ AKS Control Plane, which _improves_ extension security, efficiency, and reliability."

> ExtensionManager is moved from user node pool to AKS Control Plane, which improved extension security, efficiency, and reliability.

Reviewer: - nit: can we update 'ExtensionManager' to either 'extensions-manager' (if that's how it still appears on the cluster, I think we use 'extensions-operator' now?) or 'Extensions manager' (for a doc friendly wording)? 
- might be good to add 'AKS control plane hosted on Microsoft tenant and subscriptions'
- can we double click on how it improves security, efficiency, and reliability?

Author: - Clarified ExtensionManager and ExtensionManager pods (extension-operator and extension-agent)
- good point. Updated.
- The paragraph below explains the benefits in detail.

What Is ExtensionManager?

ExtensionManager is a key component that handles lifecycle operations (create, upgrade, delete, reconcile) for Kubernetes extensions. It periodically polls for version updates on extension instances and reconciles extension configuration settings, ensuring that extensions are managed effectively. Previously running on customer worker nodes, it has now been moved to the AKS control plane.

What Has Changed?

> Previously, extension-operator and extension-agent were deployed in the kube-system namespace by ExtensionManager on worker nodes. These components are now part of the AKS control plane, so they are no longer visible in the kube-system namespace.

Reviewer 3: I'd stick to either calling the pods out directly (`extension-operator` and `extension-agent`) or calling it ExtensionManager. Don't switch between the two, because it's sorta confusing.

Reviewer 3: i.e. here you could say "previously, ExtensionManager pods were deployed in the kube-system namespace.

> ExtensionManager is a key component that handles lifecycle operations (create, upgrade, delete, reconcile) for Kubernetes extensions. It periodically polls for version updates on extension instances and reconciles extension configuration settings, ensuring that extensions are managed effectively. Previously running on customer worker nodes, it has now been moved to the AKS control plane.

Reviewer: - nit: periodically polls which endpoint? in addition to updating the description, might be good to also link to conceptual doc on extensions where the design (along with diagram) already covers this topic

Author: Added a link to the architecture diagram.

Example of previous deployment:
```bash
$ kubectl get pods -n kube-system
NAME                                      READY   STATUS    RESTARTS   AGE
coredns-558bd4d5db-7x5c8                  1/1     Running   0          5d
extension-agent-abc123                    1/1     Running   0          3d
extension-operator-def456                 1/1     Running   0          3d

> ```

Reviewer 6: Could consider reordering these sections. From what can user see before, what changed, and what are the benefits

Benefits of the Move
1. **Improved extension integrity:** Prevents accidental interference with extension components, safeguarding extension integrity.

> 1. **Simplified Networking:** Reduces the need for complex networking configurations (as outlined in [Required FQDN / application rules](https://learn.microsoft.com/en-us/azure/aks/outbound-rules-control-egress#required-fqdn--application-rules-5)) for ExtensionManager. You only need to configure networking for the extensions themselves (e.g., allowing access to MCR for pulling container images).

Reviewer 6: You only need to configure networking for the extensions themselves as opposed to what? What action is removed because of this?

Author: As opposed to the need to configure networking for ExtensionManager. Updated for clarification.

1. **Reduced Identity Footprint:** Eliminates the need for node-level identities for ExtensionManager, improving security and reducing setup time. This also contributes to shortening the first extension installation time.

> 1. **Reduced resource usage**. Moving the ExtensionManager pods to AKS internal infrastructure will reduce the resource usage on the customer's side.

Reviewer 6: I would recommend putting this first, this is direct/visible benefit to the customer

Author: Moved the last two benefits upfront.

---
title: "Introducing Core Kubernetes Extensions for AKS"
description: "Learn what are core Kubernetes extensions, and how that extends functionality of your AKS clusters"
date: 2025-05-14 # date is important. future dates will not be published
author: Jane Guo # must match the authors.yml in the _data folder
categories: 
- add-ons-and-extensions
# - general
# - operations
# - networking
# - security
# - developer
# - add-ons-and-extensions
---

## Introducing Core Kubernetes Extensions

> ### What Are Kubernetes Extensions?

Reviewer 3: ```suggestion
### What are Kubernetes Extensions?
```

I think?

[Kubernetes extensions](https://learn.microsoft.com/azure/aks/cluster-extensions?tabs=azure-cli) (or cluster extensions) are pre-packaged applications that simplify the installation and lifecycle management of Azure capabilities on Kubernetes clusters. Examples include [Azure Backup](https://learn.microsoft.com/azure/backup/azure-kubernetes-service-backup-overview), [GitOps (Flux)](https://learn.microsoft.com/azure/azure-arc/kubernetes/conceptual-gitops-flux2), and [Azure Machine Learning](https://learn.microsoft.com/azure/machine-learning/how-to-attach-kubernetes-anywhere?view=azureml-api-2). Third-party extensions (or Kubernetes apps), such as [Datadog AKS Cluster Extension](https://azuremarketplace.microsoft.com/marketplace/apps/datadog1591740804488.dd_aks_extension?tab=Overview) and [Isovalent Cilium Enterprise](https://azuremarketplace.microsoft.com/marketplace/apps/isovalentinc1662143158090.isovalent-cilium-enterprise?tab=Overview), are also available in the [Azure Marketplace](https://azuremarketplace.microsoft.com).

### Extensions vs. Add-ons

Extensions and add-ons both enhance AKS functionality but differ in scope and management:

> - **Add-ons**: Part of the AKS [resource provider](https://learn.microsoft.com/azure/azure-resource-manager/management/resource-providers-and-types) (`Microsoft.ContainerService/managedClusters`), limited to AKS, and generally reserved for cluster critical functionality.

Reviewer 3: ```suggestion
- **Add-ons**: Part of the AKS [resource provider](https://learn.microsoft.com/azure/azure-resource-manager/management/resource-providers-and-types) (`Microsoft.ContainerService/managedClusters`), limited to AKS, and generally reserved for cluster critical functionality. These are configured directly via the Managed Cluster API.
```

Not 100% sure this is worth adding, but might be.

Reviewer: it is worth adding as we need to capture the world before core extensions for first half of the blog when we are doing context setting. essentially we need to mention that add-ons were natively available in AKS API and extensions weren't. in addition to mentioning that we need to solidify why that distinction was a problem before with the implications and pros/cons of the before and after world later in the blog

- **Extensions**: Managed via the [Extension API](https://learn.microsoft.com/rest/api/kubernetesconfiguration/extensions/extensions?view=rest-kubernetesconfiguration-extensions-2024-11-01) (`Microsoft.KubernetesConfiguration/extensions`), supporting both AKS and [Arc-enabled Kubernetes environments](https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/overview). Learn how extensions also work on Azure Arc-enabled clusters at [extension architecture](https://learn.microsoft.com/azure/azure-arc/kubernetes/conceptual-extensions#architecture).

> ### What Are Core Kubernetes Extensions?

Reviewer 3: ```suggestion
### What are Core Kubernetes Extensions?
```

> Core Kubernetes extensions are a Kubernetes extension that provides broader region availability, a native AKS experience, safer version management, enhanced security, efficiency, and reliability on AKS. Important AKS capabilities can be supported through core Kubernetes extensions while maintaining seamless integration with AKS just like add-ons. This approach also paves the way for more functionality to be supported across both AKS and Azure Arc-enabled Kubernetes environments in the future. 

Reviewer 3: ```suggestion
A Core Kubernetes extension is a Kubernetes extension that provide broader region availability, a native AKS experience, safer version management, and enhanced security, efficiency, and reliability on AKS. Important AKS capabilities can be supported through core Kubernetes extensions while maintaining seamless integration with AKS just like add-ons. This approach also paves the way for more functionality to be supported across both AKS and Azure Arc-enabled Kubernetes environments in the future. 
```

Reviewer 3: We could also rephrase "Important AKS capabilities can be supported through core Kubernetes extensions while maintaining seamless integration with AKS just like add-ons. This approach also paves the way for more functionality to be supported across both AKS and Azure Arc-enabled Kubernetes environments in the future."

to be a bit more obvious about what the users should expect?

Maybe: "Going forward, some important AKS capabilities will be offered via core extensions rather than via Addons as has been done in the past. This approach also paves the way for more functionality to be supported across both AKS and Azure Arc-enabled Kubernetes environments in the future. "

or something like that?

Reviewer 3: Helps hook the users on why they might care (because there are going to be more going forward).

> Core Kubernetes extensions are a Kubernetes extension that provides broader region availability, a native AKS experience, safer version management, enhanced security, efficiency, and reliability on AKS. Important AKS capabilities can be supported through core Kubernetes extensions while maintaining seamless integration with AKS just like add-ons. This approach also paves the way for more functionality to be supported across both AKS and Azure Arc-enabled Kubernetes environments in the future. 

Reviewer 3: Might be worth a pass to determine if you're casing `Extensions` consistently. I see some extensions and some Extensions.

#### Broader region availability

> Core Kubernetes extensions are available in all regions across Azure Public cloud, Azure Government and China regions. See the full list of regions at [Product availability by region](https://azure.microsoft.com/explore/global-infrastructure/products-by-region/table) (search "Azure Arc enabled Kubernetes" for core Kubernetes extensions and "Azure Kubernetes Service (AKS)" for AKS in the list).

Reviewer 3: The instruction here is a bit hard to parse.
> Azure Arc enabled Kubernetes

Shows ARC clusters, not extensions (but the docs imply it shows extensions). The actual region list for ARC is missing a bunch of regions there, isn't it?

Author: Yeah, agreed that it can be confusing...We don't have a region list for extensions, and that's why I just put a link to Arc regions here. I'm gonna remove this to avoid confusion.

Author: @RamyasreeChakka , @ponatara Is there a public-facing region list dedicated to extensions? Could you please verify the accuracy of the regions listed in https://azure.microsoft.com/explore/global-infrastructure/products-by-region/table?

#### Native AKS experience
##### Creating and deleting core extensions

> * **Unchanged experience for core extensions migrated from add-ons**: Some AKS add-ons, such as [Container Insights](https://learn.microsoft.com/azure/azure-monitor/containers/container-insights-overview), [Managed Prometheus](https://learn.microsoft.com/azure/azure-monitor/metrics/prometheus-metrics-overview), and [Application Monitoring](https://learn.microsoft.com/azure/azure-monitor/app/kubernetes-codeless), are transitioning to core extensions. Although the underlying technology shifts to extensions, the API and the user experience in Azure CLI, Azure Portal and ARM/Bicep/Terraform/other templates remain unchanged.

Reviewer 3: We haven't talked publicly about migrating addons to core extensions, have we?

This is the first intro to it, but it feels like a weird spot to introduce them to it?

Should we even have this here or can we cut this and put it in a different blog later?

Author: Makes sense. I'll cut it.

    CLI commands:
    * Create/delete a **standard** extension:
    ```bash
    az k8s-extension create/delete --extension-type <type> --resource-group <group> --cluster-name <name> --cluster-type <clusterType> --name <extension name>
    ```
    * Create/delete a **core** extension:
    ```bash
    az aks extension create/delete --extension-type <type> --resource-group <group> --cluster-name <name> --name <core extension name>
    ```

#### Safer version management
Core extensions follow the same [update rules as add-ons](https://learn.microsoft.com/azure/aks/integrations#add-ons): patch versions can be upgraded within a Kubernetes minor version, while major/minor upgrades occur only with Kubernetes minor version updates to avoid introducing breaking changes within a Kubernetes minor version. This policy balances reliability with the speed of availability of new features and patches.

#### Improved security, efficiency and reliability

> Extension-manager is moved from user node pool to AKS Control Plane, which improves extension security, efficiency, and reliability.

Reviewer 3: ```suggestion
Extension-manager is moved from the user node pool to the AKS Control Plane, which improves extension security, efficiency, and reliability.
```

Reviewer 3: Did we already talk about the networking advantages of this move elsewhere? If not, worth spelling them out here?

What Is extension-manager?

Extension-manager is a key component that handles lifecycle operations (create, upgrade, delete, reconcile) for Kubernetes extensions. It periodically polls for version updates on extension instances and reconciles extension configuration settings, ensuring that extensions are managed effectively. Learn more about the flow in [extension architecture](https://learn.microsoft.com/azure/azure-arc/kubernetes/conceptual-extensions#architecture). 

Previously, extension-manager pods (extension-operator and extension-agent) were deployed in the kube-system namespace on customer worker nodes. Example of a deployment before:
```bash
$ kubectl get pods -n kube-system
NAME                                      READY   STATUS    RESTARTS   AGE
coredns-558bd4d5db-7x5c8                  1/1     Running   0          5d
extension-agent-abc123                    1/1     Running   0          3d
extension-operator-def456                 1/1     Running   0          3d
```

> What Has Changed?

Reviewer 3: These look like headings -- maybe bold them?

Extension-manager components are now moved to the AKS control plane hosted on Microsoft tenant and subscriptions. Extension-manager is no longer visible to users.

Benefits of the Move

> 1. **Reduced Identity Footprint:** Eliminates the need for node-level identities for extension-manager, improving security and reducing setup time. This also contributes to shortening the first extension installation time.

Reviewer: if we are mentioning node level identity then we need to capture which identity specifically was extension manager using before?

Author: Added a link to extension managed identity requirements

Author: Updated.

1. **Reduced resource usage**. Moving the extension-manager pods to AKS internal infrastructure will reduce the resource usage on the customer's side.
1. **Improved extension integrity:** Prevents accidental interference with extension components, safeguarding extension integrity.

> 1. **Simplified Networking:** Reduces the need for complex networking configurations. Before the migration, users need to set up required networking rules as outlined in [Required FQDN / application rules](https://learn.microsoft.com/en-us/azure/aks/outbound-rules-control-egress#required-fqdn--application-rules-5) for extension-manager. Extension-manager now communicates directly with the following endpoints from AKS Control Plane, removing the need for users to configure them. Note that some extensions may still require users to configure networking for the extension instances on the users' clusters (e.g., allowing access to MCR for pulling container images).

Reviewer: we need to add a line here that this move was a blocker before for network isolated clusters. and that with this move, one of the blockers to extensions working on network isolated clusters is now removed. @janeziqiguo would recommend syncing with @karataliu. above removes network isolation hurdle. I think there might still be another open item on syncing extension charts and images to bootstrap ACR still pending

    * `<region>.dp.kubernetesconfiguration.azure.com`
    * `mcr.microsoft.com`, `*.data.mcr.microsoft.com`
    * `arcmktplaceprod.azurecr.io` and `arcmktplaceprod.<region>.data.azurecr.io`

## Which applications or services will be supported through core Kubernetes extensions?

> We're in the process of enabling Container insights, Managed prometheus, App monitoring, and Azure Backup through core extensions, and there will be more. Keep an eye out for more updates, and share your feedback with us!

Reviewer 3: Do we have links we can put here for these? (I think you had some above, maybe proxy down here too?)
